File: app/javascript/channels/rating_channel.js
import consumer from "./consumer"

// given the nesting, capturing this resolves access problems.
const ratingChannel = consumer.subscriptions.create({ channel: "RatingChannel"},
{
  initialized() {
    this.installOnAllItems();
  },

  received(data) {
    if (data['command'] == 'replace') {
      this.replaceItem(data);
    } else {
      // leave room for other commands, but for now just choke
      console.log("UNKNOWN COMMAND: " + data['command']);
    }
  },

  replaceItem(data) {
    // replace the contents of the container
    const item = document.getElementById("item_" + data['item_id'])
    item.innerHTML = data['html'];
    ratingChannel.installOnItem(item);
  },

  incrementItem() {
    ratingChannel.perform('inc_rating', {'item_id': this.dataset.itemId});
  },

  decrementItem() {
    ratingChannel.perform('dec_rating', {'item_id': this.dataset.itemId});
  },

  installOnAllItems() {
    document.addEventListener("DOMContentLoaded", (event) => {
      const items = Array.from(document.getElementsByClassName('item-row'));
      items.forEach(item => {
        this.installOnItem(item);
      });
    });  
  },

  installOnItem(item) {
    // <button class="item-increment" ...>
    let incButton = item.getElementsByClassName('item-increment')[0];
    let decButton = item.getElementsByClassName('item-decrement')[0];

    incButton.addEventListener('click', this.incrementItem);
    decButton.addEventListener('click', this.decrementItem);
  }
});

File: app/channels/rating_channel.rb
class RatingChannel < ApplicationCable::Channel
  def subscribed
    # this isn't tied to a model or the user. it is 
    # a universal broadcast
    stream_from "rating"
  end

  def receive(data)
    ActionCable.server.broadcast("rating", data)
  end

  def inc_rating(data)
    broadcast_item(Item.find_by_id(data['item_id']).increment!(:rating))
  end

  def dec_rating(data)
    broadcast_item(Item.find_by_id(data['item_id']).decrement!(:rating))
  end

  private

  def broadcast_item(item)
    html = ApplicationController
            .renderer
            .render_to_string(partial: 'items/item_template',
                              locals: { item: item })
    ActionCable.server.broadcast("rating", {html: html, item_id: item.id, command: "replace"})
  end
end